lib/mtimes.js

```diff
@@ -0,0 +1,68 @@
+'use strict';
+
+const fs = require('graceful-fs');
+
+const MTIME = new Date(1552815428391);
+
+function getPrecision(path, options, callback) {
+    options = options || {};
+    options.fs = options.fs || fs;
+
+    options.fs.stat(path, (err, stat) => {
+        if (err) {
+            return callback(err);
+        }
+
+        // Attempt to update mtime to our known past value
+        options.fs.utimes(path, stat.atime, MTIME, (err) => {
+            if (err) {
+                return callback(err);
+            }
+
+            // Read back the mtime and see if it is still ms precision
+            options.fs.stat(path, (err, updatedStat) => {
+                if (err) {
+                    return callback(err);
+                }
+
+                // Put back the original mtime, nothing to see here!
+                options.fs.utimes(path, stat.atime, stat.mtime, (err) => {
+                    if (err) {
+                        return callback(err);
+                    }
+
+                    if (updatedStat.mtime.getTime() === MTIME.getTime()) {
+                        callback(null, 'ms');
+                    } else if (updatedStat.mtime.getTime() === (Math.trunc(MTIME.getTime() / 1000) * 1000)) {
+                        callback(null, 's');
+                    } else {
+                        callback(Object.assign(
+                            new Error('Unsupported mtime precision'),
+                            { code: 'EMTIMEPRECISION' }
+                        ));
+                    }
+                });
+            });
+        });
+    });
+}
+
+// Determine if mtimes are equal according to precision
+function isEqual(t0, t1, precision) {
+    if (t0.getTime() === t1.getTime()) {
+        return true;
+    }
+
+    if (precision === 'ms') {
+        return false;
+    }
+
+    if (precision === 's') {
+        return Math.trunc(t0.getTime() / 1000) === Math.trunc(t1.getTime() / 1000);
```

Comment:
A concern I have with this approach is that it makes assumptions about how the os is handling precision. Are the numbers always truncated, or is it possible for rounding take place? If it's possible for rounding to take place, we might hit occasional failures because we're comparing against a truncated value. If we always handle the truncation up front and set `utimes` anytime the lock is created/updated with an already truncated number, we should be able to avoid any potential rounding issues.

Reply:
@jacobheun that's a valid concern, meaning we should truncate or not truncate `Date.now` based on the calculated precision. This would solve the issue.

Reply:
@satazor with the updated implementation we don't necessarily know the precision when we use `utimes`...

Reply:
So, I had a chat with @alanshaw and there was two alternatives:

1. Keep using the `Date.now` in ms and deal with possible rounding issues when checking
2. Give a `Date.now` output based on the precision being calculated, like:

```js
const createMtime = () => {
    let precision = 's';

    return {
        analyze: (statMtime) => {
            // Detect `statMtime` precision; this should be called after all stat calls, right after mkdir and when stating in the updateLock
        },
        now: () => {
            // Return now based on the calculated precision
        }
    }
}
```

The current PR has the `1` approach, which is somewhat simpler, though I find the `2` approach more correct as there's no rounding issues to deal with.

I'm fine with merging the PR has is though, since we are in a hurry :D

Reply:
I think merging this and doing a followup with 2 is a reasonable path forward. The current approach should account for most cases, but approach 2 is better in the long run to avoid edge cases.

Reply:
Merging it as is now and I will create an issue for it.